File: algorithms/cpp/_069_Sqrt_x/_069_Sqrt_x_std.cpp
#include <iostream>

using namespace std;

class Solution {
public:
    /**
     * @param x: An integer
     * @return: The sqrt of x
     */
    int sqrt(int x) {
        // write your code here
        if (0 == x) {
            return 0;
        }
        int left = 1;
        int right = x / 2 + 1;
        while (left + 1 < right) {
            long long mid = left + (right - left) / 2;
            long long multiply = mid * mid;
            //cout << multiply << " " << left << " " << right << endl;
            if (multiply == x) {
                return mid;
            }
            if (multiply > x) {
                right = mid;
            } else {
                left = mid;
            }
        }
        if ((long long) (right) * right < x) {
            return right;
        }
        return left;
    }
};

int main() {
    auto *so = new Solution();
    cout << "new file!" << endl;
    return 0;
}

File: algorithms/java/src/_069_Sqrt_x/_069_Sqrt_x_std.java
package _069_Sqrt_x;

public class _069_Sqrt_x_std {
    public static void main(String[] args) {
        _069_Sqrt_x_std_ so = new _069_Sqrt_x_std_();

    }
}

class _069_Sqrt_x_std_ {

}

class Solution {
    /**
     * @param x: An integer
     * @return: The sqrt of x
     */

    public int sqrt(int x) {
        int start = 0;
        int end = x;
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            int div = x / mid;
            if (div > mid) {
                start = mid;
            } else if (div == mid) {
                return mid;
            } else {
                end = mid;
            }
        }
        if (x != 0 && x / end >= end) {
            return end;
        } else {
            return start;
        }
    }
}